File: sources/iOS/ios-communications/iOSCommunicationsTests/PolarAdvDataUtilityTest.swift
//  Copyright © 2021 Polar. All rights reserved.

import XCTest
import iOSCommunications

class PolarAdvDataUtilityTest: XCTestCase {
    
    func testGetPolarModelNameFromAdvLocalNameOneWord() throws {
        // Arrange
        let advName = "Polar OneWord1234 12345678"
        let modelNameExpected = "OneWord1234"
        
        // Act
        let modelName = PolarAdvDataUtility.getPolarModelNameFromAdvLocalName(advLocalName: advName)
        
        // Assert
        XCTAssertEqual(modelNameExpected, modelName)
    }
    
    func testGetPolarModelNameFromAdvLocalNameTwoWords() throws {
        // Arrange
        let advName = " Polar two woRds 12345678 "
        let modelNameExpected = "two woRds"
        
        // Act
        let modelName = PolarAdvDataUtility.getPolarModelNameFromAdvLocalName(advLocalName: advName)
        
        // Assert
        XCTAssertEqual(modelNameExpected, modelName)
    }
    
    func testGetPolarModelNameFromAdvLocalNameMultiWords() throws {
        // Arrange
        let advName = "Polar Some m x name 12345678"
        let modelNameExpected = "Some m x name"
        
        // Act
        let modelName = PolarAdvDataUtility.getPolarModelNameFromAdvLocalName(advLocalName: advName)
        
        // Assert
        XCTAssertEqual(modelNameExpected, modelName)
    }
    
    func testGetPolarModelNameFromAdvLocalNameNoPolarPrefix() throws {
        // Arrange
        let advNameMultiWords = "Some m x name Polar 12345678"
        let modelNameExpected = ""
        
        // Act
        let modelName = PolarAdvDataUtility.getPolarModelNameFromAdvLocalName(advLocalName: advNameMultiWords)
        
        // Assert
        XCTAssertEqual(modelNameExpected, modelName)
    }

    func testGetPolarModelNameFromAdvLocalNameDeviceIdMissing() throws {
        // Arrange
        let advNameMultiWords = "Polar noId "
        let modelNameExpected = ""

        // Act
        let modelName = PolarAdvDataUtility.getPolarModelNameFromAdvLocalName(advLocalName: advNameMultiWords)

        // Assert
        XCTAssertEqual(modelNameExpected, modelName)
    }
    
    func testIsPolarDeviceWithPolarPrefix() throws {
        // Arrange
        let advName = " Polar Some m x name 12345678"
        
        // Act
        let isPolarDevice = PolarAdvDataUtility.isPolarDevice(advLocalName: advName)
        
        // Assert
        XCTAssertTrue(isPolarDevice)
    }
    
    func testIsPolarDeviceNoPolarPrefix() throws {
        // Arrange
        let advName = "Some m x name Polar 12345678"
        
        // Act
        let isPolarDevice = PolarAdvDataUtility.isPolarDevice(advLocalName: advName)
        
        // Assert
        XCTAssertFalse(isPolarDevice)
    }
    
    func testIsPolarDeviceOnlyPrefix() throws {
        // Arrange
        let advName = "Polar"
        
        // Act
        let isPolarDevice = PolarAdvDataUtility.isPolarDevice(advLocalName: advName)
        
        // Assert
        XCTAssertFalse(isPolarDevice)
    }
}

File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/api/ble/model/polar/PolarAdvDataUtilityTest.kt
package com.polar.androidcommunications.api.ble.model.polar

import org.junit.Assert
import org.junit.Test

class PolarAdvDataUtilityTest {

    @Test
    fun `getDeviceModelNameFromAdvLocalName() one word`() {
        // Arrange
        val advName = "Polar OneWord1234 12345678"
        val modelNameExpected = "OneWord1234"

        // Act
        val modelName = PolarAdvDataUtility.getDeviceModelNameFromAdvLocalName(advName)

        // Assert
        Assert.assertEquals(modelNameExpected, modelName)
    }

    @Test
    fun `getDeviceModelNameFromAdvLocalName() two word`() {
        // Arrange
        val advName = " Polar two woRds 12345678 "
        val modelNameExpected = "two woRds"

        // Act
        val modelName = PolarAdvDataUtility.getDeviceModelNameFromAdvLocalName(advName)

        // Assert
        Assert.assertEquals(modelNameExpected, modelName)
    }

    @Test
    fun `getDeviceModelNameFromAdvLocalName() multi word`() {
        // Arrange
        val advName = "Polar Some m x name 12345678"
        val modelNameExpected = "Some m x name"

        // Act
        val modelName = PolarAdvDataUtility.getDeviceModelNameFromAdvLocalName(advName)

        // Assert
        Assert.assertEquals(modelNameExpected, modelName)
    }

    @Test
    fun `getDeviceModelNameFromAdvLocalName() no Polar prefix`() {
        // Arrange
        val advNameMultiWords = "Some m x name Polar 12345678"
        val modelNameExpected = ""

        // Act
        val modelName = PolarAdvDataUtility.getDeviceModelNameFromAdvLocalName(advNameMultiWords)

        // Assert
        Assert.assertEquals(modelNameExpected, modelName)
    }

    @Test
    fun `getDeviceModelNameFromAdvLocalName() device id missing`() {
        // Arrange
        val advNameMultiWords = "Polar noId "
        val modelNameExpected = ""

        // Act
        val modelName = PolarAdvDataUtility.getDeviceModelNameFromAdvLocalName(advNameMultiWords)

        // Assert
        Assert.assertEquals(modelNameExpected, modelName)
    }

    @Test
    fun `isValidDevice() with Polar prefix`() {
        // Arrange
        val advName = " Polar Some m x name 12345678"

        // Act
        val isValidDevice = PolarAdvDataUtility.isValidDevice(advName)

        // Assert
        Assert.assertTrue(isValidDevice)
    }

    @Test
    fun `isValidDevice() no Polar prefix`() {
        // Arrange
        val advName = "Some m x name Polar 12345678"

        // Act
        val isValidDevice = PolarAdvDataUtility.isValidDevice(advName)

        // Assert
        Assert.assertFalse(isValidDevice)
    }

    @Test
    fun `isValidDevice() only prefix`() {
        // Arrange
        val advName = "Polar"

        // Act
        val isValidDevice = PolarAdvDataUtility.isValidDevice(advName)

        // Assert
        Assert.assertFalse(isValidDevice)
    }
}